dpc/src/circuits/inner_circuit_gadget.rs

```diff
@@ -721,386 +690,134 @@ where
-
-                let element_squared = element
-                    .0
-                    .mul(encryption_cs.ns(|| format!("element_{} ^ 2", i)), &element.0)?;
-
-                // Enforce that either a_i == element_squared or b_i == element_squared
-
-                let a_i_minus_element_squared = a_i.sub(
-                    encryption_cs.ns(|| format!("a_{} - element_{}^2", i, i)),
-                    &element_squared,
-                )?;
-                let b_i_minus_element_squared = b_i.sub(
-                    encryption_cs.ns(|| format!("b_{} - element_{}^2", i, i)),
-                    &element_squared,
-                )?;
-
-                let expected_zero = a_i_minus_element_squared.mul(
-                    encryption_cs.ns(|| format!("(-a_{} - element_{}^2) * (-b_{} - element_{}^2)", i, i, i, i)),
-                    &b_i_minus_element_squared,
-                )?;
-
-                expected_zero.enforce_equal(
-                    encryption_cs.ns(|| format!("(-a_{} - element_{}^2) * (-b_{} - element_{}^2) == 0", i, i, i, i)),
-                    &fp_zero,
-                )?;
-
-                a_i.enforce_not_equal(encryption_cs.ns(|| format!("a_{} != b_{}", i, i)), &b_i)?;
-
-                element_squared
-                    .enforce_not_equal(encryption_cs.ns(|| format!("element_squared_{} != 0", i)), &fp_zero)?;
-            }
-
             // *******************************************************************
             // Construct the record encryption
 
-            let encryption_randomness_gadget = AccountEncryptionGadget::RandomnessGadget::alloc(
+            let encryption_randomness_gadget = <C::AccountEncryptionGadget as EncryptionGadget<
+                C::AccountEncryptionScheme,
+                C::InnerScalarField,
+            >>::RandomnessGadget::alloc(
                 &mut encryption_cs.ns(|| format!("output record {} encryption_randomness", j)),
                 || Ok(encryption_randomness),
             )?;
 
-            let encryption_blinding_exponents_gadget = AccountEncryptionGadget::EncryptionWitnessGadget::alloc(
+            // Technically, an encryption scheme does not need blinding exponents.
+            //
+            // Before we remove it from the trait, we allocate them by default.
+            let encryption_blinding_exponents = vec![
+                    <C::AccountEncryptionScheme as EncryptionScheme>::EncryptionWitness::default();
+                    record_field_elements_gadgets.len()
+                ];
+
+            let encryption_blinding_exponents_gadget = <C::AccountEncryptionGadget as EncryptionGadget<
+                C::AccountEncryptionScheme,
+                C::InnerScalarField,
+            >>::EncryptionWitnessGadget::alloc(
                 &mut encryption_cs.ns(|| format!("output record {} encryption_blinding_exponents", j)),
                 || Ok(encryption_blinding_exponents),
             )?;
```

Comment:
Can you clarify for me what is happening here with the blinding exponents?

I understand that this scheme doesn't require the blinding exponents (as before), and has an encryption witness that is added and subtracted to get the plaintext. Is this technically an allocation of those witnesses?

Reply:
Yes, this is technically an allocation of dummy witnesses (for ECIES Poseidon, the encryption witness is empty).

The reason is somehow this: if one is given the randomness, the encryption key, and the plaintext, the person should be able to deduce the witnesses from it. (Therefore, technically the "witnesses" can be internalized by the gadgets and do not need a separate interface.)

Reply:
Just to check then, do we actually need to use `record_field_elements_gadgets` in this implementation anymore? Given that we used this as a shortcut in the past, I want to see if this scheme still will benefit from it (since we do pay the constraint cost of checking `expected_plaintext_bits == record_field_elements_gadgets`.

Reply:
It seems needed, as otherwise the ciphertext could contain a different plaintext.